File: bin/SingineServe.java
///usr/bin/env jbang "$0" "$@" ; exit $?
// Or: java SingineServe.java [port] [webroot]
// Requires Java 11+. No dependencies.
/**
 * SingineServe — singine's Java HTTP appserver for SilkPage documents.
 *
 * Single-file Java server using com.sun.net.httpserver (built into every JDK).
 * Serves the SilkPage www/ webroot statically.
 *
 * Usage:
 *   java bin/SingineServe.java [port] [webroot]
 *
 * Defaults:
 *   port    = 4242
 *   webroot = ../../silkpage/main/silkpage/www  (relative to singine repo root)
 *
 * Routes:
 *   GET /     → index.html
 *   GET /cv/  → cv/index.html
 *   GET /**   → static file
 */

import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpHandler;
import com.sun.net.httpserver.HttpServer;

import java.io.*;
import java.net.InetSocketAddress;
import java.net.URI;
import java.nio.charset.StandardCharsets;
import java.nio.file.*;
import java.util.Map;
import java.util.concurrent.Executors;

public class SingineServe {

    static final Map<String, String> MIME = Map.ofEntries(
        Map.entry(".html",  "text/html; charset=utf-8"),
        Map.entry(".css",   "text/css; charset=utf-8"),
        Map.entry(".js",    "application/javascript; charset=utf-8"),
        Map.entry(".mjs",   "application/javascript; charset=utf-8"),
        Map.entry(".json",  "application/json; charset=utf-8"),
        Map.entry(".xml",   "application/xml; charset=utf-8"),
        Map.entry(".xsl",   "application/xslt+xml; charset=utf-8"),
        Map.entry(".xsd",   "application/xml; charset=utf-8"),
        Map.entry(".svg",   "image/svg+xml"),
        Map.entry(".png",   "image/png"),
        Map.entry(".jpg",   "image/jpeg"),
        Map.entry(".jpeg",  "image/jpeg"),
        Map.entry(".ico",   "image/x-icon"),
        Map.entry(".pdf",   "application/pdf"),
        Map.entry(".ttf",   "font/ttf"),
        Map.entry(".woff",  "font/woff"),
        Map.entry(".woff2", "font/woff2"),
        Map.entry(".txt",   "text/plain; charset=utf-8")
    );

    public static void main(String[] args) throws Exception {
        int    port    = args.length > 0 ? Integer.parseInt(args[0]) : 4242;
        String envRoot = System.getenv("SINGINE_WEBROOT");

        Path webroot;
        if (args.length > 1) {
            webroot = Path.of(args[1]).toAbsolutePath().normalize();
        } else if (envRoot != null) {
            webroot = Path.of(envRoot).toAbsolutePath().normalize();
        } else {
            // Default: singine repo root → sibling silkpage www/
            Path binDir  = Path.of(SingineServe.class.getProtectionDomain()
                                    .getCodeSource().getLocation().toURI()).getParent();
            Path repoRoot = binDir.getParent();          // singine/
            webroot = repoRoot.resolve("../../silkpage/main/silkpage/www")
                              .toAbsolutePath().normalize();
        }

        if (!Files.isDirectory(webroot)) {
            System.err.println("singine-serve: webroot not found: " + webroot);
            System.err.println("  Set SINGINE_WEBROOT or pass as second argument.");
            System.exit(1);
        }

        final Path root = webroot;
        HttpServer server = HttpServer.create(new InetSocketAddress("127.0.0.1", port), 0);
        server.createContext("/", new StaticHandler(root));
        server.setExecutor(Executors.newCachedThreadPool());
        server.start();

        System.out.println("singine-serve (Java) running");
        System.out.println("  webroot  " + root);
        System.out.println("  url      http://localhost:" + port + "/");
        System.out.println("  cv       http://localhost:" + port + "/cv/");
        System.out.println();
        System.out.println("  Ctrl+C to stop");
    }

    static class StaticHandler implements HttpHandler {
        private final Path root;
        StaticHandler(Path root) { this.root = root; }

        @Override
        public void handle(HttpExchange ex) throws IOException {
            if (!"GET".equalsIgnoreCase(ex.getRequestMethod()) &&
                !"HEAD".equalsIgnoreCase(ex.getRequestMethod())) {
                send(ex, 405, "text/plain", "405 Method Not Allowed".getBytes());
                return;
            }

            String rawPath = ex.getRequestURI().getPath();
            if (rawPath == null || rawPath.isEmpty()) rawPath = "/";

            // Decode and normalise
            String decoded = URI.create(rawPath).getPath();
            if (decoded.endsWith("/")) decoded += "index.html";

            Path abs = root.resolve("." + decoded).normalize();

            // Reject path traversal
            if (!abs.startsWith(root)) {
                send(ex, 403, "text/plain", "403 Forbidden".getBytes());
                return;
            }

            // If it's a directory, try index.html
            if (Files.isDirectory(abs)) abs = abs.resolve("index.html");

            if (!Files.isRegularFile(abs)) {
                byte[] body = ("404 Not Found: " + decoded).getBytes(StandardCharsets.UTF_8);
                send(ex, 404, "text/plain; charset=utf-8", body);
                return;
            }

            String ext  = ext(abs.getFileName().toString());
            String mime = MIME.getOrDefault(ext, "application/octet-stream");
            byte[] body = Files.readAllBytes(abs);

            ex.getResponseHeaders().set("Content-Type", mime);
            ex.sendResponseHeaders(200, body.length);
            if (!"HEAD".equalsIgnoreCase(ex.getRequestMethod())) {
                try (OutputStream os = ex.getResponseBody()) { os.write(body); }
            } else {
                ex.getResponseBody().close();
            }
        }

        private static void send(HttpExchange ex, int code, String mime, byte[] body)
                throws IOException {
            ex.getResponseHeaders().set("Content-Type", mime);
            ex.sendResponseHeaders(code, body.length);
            try (OutputStream os = ex.getResponseBody()) { os.write(body); }
        }

        private static String ext(String filename) {
            int i = filename.lastIndexOf('.');
            return i < 0 ? "" : filename.substring(i).toLowerCase();
        }
    }
}

File: bin/singine-serve.mjs
#!/usr/bin/env node
/**
 * singine-serve — lightweight HTTP appserver for SilkPage documents.
 *
 * Serves the SilkPage www/ webroot via singine's execution engine.
 * Follows the same Node.js + ESM pattern as singine-mcp.
 *
 * Usage:
 *   node bin/singine-serve.mjs [port] [webroot]
 *   singine serve [port]           # via singine CLI shim (future)
 *
 * Defaults:
 *   port    = 4242
 *   webroot = ../../silkpage/main/silkpage/www  (relative to singine repo root)
 *
 * Routes:
 *   GET /          → webroot/index.html
 *   GET /cv/       → webroot/cv/index.html    ← CV preview
 *   GET /**        → static file from webroot
 *
 * MIME types handled: html, css, js, mjs, json, xml, xsl, svg,
 *                     png, jpg, gif, webp, ico, pdf, ttf, woff2
 */

import http    from 'node:http';
import fs      from 'node:fs';
import path    from 'node:path';
import { URL } from 'node:url';

// ── Config ─────────────────────────────────────────────────────────────────

const PORT    = parseInt(process.argv[2] ?? process.env.SINGINE_PORT ?? '4242', 10);
const __dir   = path.dirname(new URL(import.meta.url).pathname);
const REPO    = path.resolve(__dir, '..');
const WEBROOT = process.argv[3]
  ?? process.env.SINGINE_WEBROOT
  ?? path.resolve(REPO, '../../silkpage/main/silkpage/www');

const MIME = {
  '.html': 'text/html; charset=utf-8',
  '.css':  'text/css; charset=utf-8',
  '.js':   'application/javascript; charset=utf-8',
  '.mjs':  'application/javascript; charset=utf-8',
  '.json': 'application/json; charset=utf-8',
  '.xml':  'application/xml; charset=utf-8',
  '.xsl':  'application/xslt+xml; charset=utf-8',
  '.xsd':  'application/xml; charset=utf-8',
  '.svg':  'image/svg+xml',
  '.png':  'image/png',
  '.jpg':  'image/jpeg',
  '.jpeg': 'image/jpeg',
  '.gif':  'image/gif',
  '.webp': 'image/webp',
  '.ico':  'image/x-icon',
  '.pdf':  'application/pdf',
  '.ttf':  'font/ttf',
  '.woff': 'font/woff',
  '.woff2':'font/woff2',
};

// ── Request handler ─────────────────────────────────────────────────────────

function handler(req, res) {
  const url      = new URL(req.url, `http://localhost:${PORT}`);
  let   pathname = decodeURIComponent(url.pathname);

  // Directory → index.html
  if (pathname.endsWith('/')) pathname += 'index.html';

  // Resolve against webroot; reject path traversal
  const abs = path.resolve(WEBROOT, '.' + pathname);
  if (!abs.startsWith(WEBROOT)) {
    res.writeHead(403); res.end('403 Forbidden'); return;
  }

  fs.stat(abs, (err, stat) => {
    if (err || !stat.isFile()) {
      // Try appending index.html for bare directory paths without trailing slash
      const idx = path.join(abs, 'index.html');
      fs.stat(idx, (e2, s2) => {
        if (!e2 && s2.isFile()) return sendFile(idx, res);
        res.writeHead(404, { 'Content-Type': 'text/plain' });
        res.end(`404 Not Found: ${pathname}`);
      });
      return;
    }
    sendFile(abs, res);
  });
}

function sendFile(abs, res) {
  const ext  = path.extname(abs).toLowerCase();
  const mime = MIME[ext] ?? 'application/octet-stream';
  const stream = fs.createReadStream(abs);
  stream.on('error', () => { res.writeHead(500); res.end('500 Read error'); });
  res.writeHead(200, { 'Content-Type': mime });
  stream.pipe(res);
}

// ── Start ───────────────────────────────────────────────────────────────────

if (!fs.existsSync(WEBROOT)) {
  console.error(`singine-serve: webroot not found: ${WEBROOT}`);
  console.error(`  Set SINGINE_WEBROOT or pass as second argument.`);
  process.exit(1);
}

const server = http.createServer(handler);
server.listen(PORT, '127.0.0.1', () => {
  console.log(`singine-serve running`);
  console.log(`  webroot  ${WEBROOT}`);
  console.log(`  url      http://localhost:${PORT}/`);
  console.log(`  cv       http://localhost:${PORT}/cv/`);
  console.log();
  console.log(`  Ctrl+C to stop`);
});
